haystack/core/pipeline/pipeline.py

```diff
@@ -71,191 +172,220 @@ def _run_component(
-                        if receiver in cycle:
-                            final_output_reached = True
-                            break
-                    if final_output_reached:
-                        break
-
-                if not final_output_reached:
-                    # We stop only if the Component we just ran doesn't send any output to sockets that
-                    # are part of the cycle
-                    cycle_received_inputs = True
-
-                # We manage to run this component that was in the waiting list, we can remove it.
-                # This happens when a component was put in the waiting list but we reached it from another edge.
-                _dequeue_waiting_component((name, comp), waiting_queue)
-                for pair in self._find_components_that_will_receive_no_input(name, res, components_inputs):
-                    _dequeue_component(pair, run_queue, waiting_queue)
-
-                receivers = [item for item in self._find_receivers_from(name) if item[0] in cycle]
-
-                res = self._distribute_output(receivers, res, components_inputs, run_queue, waiting_queue)
-
-                # We treat a cycle as a completely independent graph, so we keep track of output
-                # that is not sent inside the cycle.
-                # This output is going to get distributed to the wider graph after we finish running
-                # a cycle.
-                # All values that are left at this point go outside the cycle.
-                if len(res) > 0:
-                    subgraph_outputs[name] = res
+        Calculates the execution priority for each component and inserts it into the priority queue.
+
+        :param component_names: Names of the components to put into the queue.
+        :param inputs: Inputs to the components.
+        :returns: A prioritized queue of component names.
+        """
+        priority_queue = FIFOPriorityQueue()
+        for component_name in component_names:
+            component = self._get_component_with_graph_metadata(component_name)
+            priority = self._calculate_priority(component, inputs.get(component_name, {}))
+            priority_queue.push(component_name, priority)
+
+        return priority_queue
+
+    @staticmethod
+    def _calculate_priority(component: Dict, inputs: Dict) -> ComponentPriority:
+        """
+        Calculates the execution priority for a component depending on the component's inputs.
+
+        :param component: Component metadata and component instance.
+        :param inputs: Inputs to the component.
+        :returns: Priority value for the component.
+        """
+        if not can_component_run(component, inputs):
+            return ComponentPriority.BLOCKED
+        elif is_any_greedy_socket_ready(component, inputs) and are_all_sockets_ready(component, inputs):
+            return ComponentPriority.HIGHEST
```

Comment:
@mathislucka, could you help clarify our goal behind assigning “highest priority” to components?

If `HIGHEST` means running as soon as all mandatory sockets are ready and at least one greedy socket has inputs, then we should consider setting `only_check_mandatory=True`. I thought about this because as per my understanding a “greedy” socket implies the component is willing to run multiple times or run as soon as partial input is available. Is the goal that we don’t block execution due to optional sockets that haven’t yet received inputs?
```suggestion
        elif is_any_greedy_socket_ready(component, inputs) and are_all_sockets_ready(component, inputs, only_check_mandatory=True):
            return ComponentPriority.HIGHEST
```

Reply:
Yes, I initially did not wait for optional inputs but I checked the existing logic and we wait for optional inputs there, so I did not want to change that.

I think that in practice we don't have any components that receive both a greedy variadic and an optional input.